controllers/topology/reconcile_state_test.go

```diff
@@ -203,6 +203,259 @@ func TestReconcileInfrastructureCluster(t *testing.T) {
+	md7 := newFakeMachineDeploymentTopologyState("md-7", infrastructureMachineTemplate7, bootstrapTemplate7)
+
+	tests := []struct {
+		name                                      string
+		current                                   []machineDeploymentTopologyState
+		desired                                   []machineDeploymentTopologyState
+		want                                      []machineDeploymentTopologyState
+		wantInfrastructureMachineTemplateRotation map[string]bool
+		wantBootstrapTemplateRotation             map[string]bool
+		wantErr                                   bool
+	}{
+		{
+			name:    "Should create desired MachineDeployment if the current does not exists yet",
+			current: nil,
+			desired: []machineDeploymentTopologyState{md1},
+			want:    []machineDeploymentTopologyState{md1},
+			wantErr: false,
+		},
+		{
+			name:    "No-op if current MachineDeployment is equal to desired",
+			current: []machineDeploymentTopologyState{md1},
+			desired: []machineDeploymentTopologyState{md1},
+			want:    []machineDeploymentTopologyState{md1},
+			wantErr: false,
+		},
+		{
+			name:    "Should update MachineDeployment with InfrastructureMachineTemplate rotation",
+			current: []machineDeploymentTopologyState{md2},
+			desired: []machineDeploymentTopologyState{md2WithRotatedInfrastructureMachineTemplate},
+			want:    []machineDeploymentTopologyState{md2WithRotatedInfrastructureMachineTemplate},
+			wantInfrastructureMachineTemplateRotation: map[string]bool{"md-2": true},
+			wantErr: false,
+		},
+		{
+			name:                          "Should update MachineDeployment with BootstrapTemplate rotation",
+			current:                       []machineDeploymentTopologyState{md3},
+			desired:                       []machineDeploymentTopologyState{md3WithRotatedBootstrapTemplate},
+			want:                          []machineDeploymentTopologyState{md3WithRotatedBootstrapTemplate},
+			wantBootstrapTemplateRotation: map[string]bool{"md-3": true},
+			wantErr:                       false,
+		},
+		{
+			name:    "Should update MachineDeployment with InfrastructureMachineTemplate and BootstrapTemplate rotation",
+			current: []machineDeploymentTopologyState{md4},
+			desired: []machineDeploymentTopologyState{md4WithRotatedTemplates},
+			want:    []machineDeploymentTopologyState{md4WithRotatedTemplates},
+			wantInfrastructureMachineTemplateRotation: map[string]bool{"md-4": true},
+			wantBootstrapTemplateRotation:             map[string]bool{"md-4": true},
+			wantErr:                                   false,
+		},
+		{
+			name:    "Should fail update MachineDeployment because of changed InfrastructureMachineTemplate kind",
+			current: []machineDeploymentTopologyState{md5},
+			desired: []machineDeploymentTopologyState{md5WithChangedInfrastructureMachineTemplateKind},
+			wantErr: true,
+		},
+		{
+			name:    "Should fail update MachineDeployment because of changed BootstrapTemplate namespace",
```

Comment:
Note, this should never happen due to webhooks, I would remove this test

Reply:
My reasoning was that when we implement the check in our "prod" code (event though this situation shouldn't happen) it would also be good to have it not untested. In case you still want to delete it, let me know, it would be fine for me.

Reply:
it is ok to have the test